[Remote Inspection] Add the ability to fetch all targetable elements

aspensantos opened the pull request:
#### da7b72f3553442d570ba82fdcb4f187911c80999
<pre>
[Remote Inspection] Add the ability to fetch all targetable elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=281173">https://bugs.webkit.org/show_bug.cgi?id=281173</a>
<a href="https://rdar.apple.com/135354335">rdar://135354335</a>

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Add the ability to fetch all targetable elements to collect highlight regions for accessibility

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::topologicallySortElementsHelper):
(WebCore::ElementTargetingController::topologicallySortElements):
- Topologically sort elements represented in a directed graph

(WebCore::ElementTargetingController::findAllTargets):
- Find all the targets in view by using the hitTestInterval parameter to findTargets in sections of that size
- Store the z relationships between the targets found in the dictionary elementIDToOccludedElementIDs
- Use the topologicallySortElements function on elementIDToOccludedElementIDs to get the z sorted elementIDs
- Return the targets for each elementID in z sorted order

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestAllTargetableElementsInfo:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestAllTargetableElements):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestAllTargetableElements):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(-[WKWebView allTargetableElementsWithHitTestInterval:]):
(TestWebKitAPI::TEST(ElementTargeting, RequestAllVisibleElements)):
- Add API test for [WKWebView _requestAllTargetableElementsInfo:completionHandler:]

* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-11.html: Added.
- Add html file used in API test for [WKWebView _requestAllTargetableElementsInfo:completionHandler:]

Canonical link: <a href="https://commits.webkit.org/285200@main">https://commits.webkit.org/285200@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/484b3ea162fb061f81ec13bcb33d8394873e355e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71732 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51145 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75847 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22937 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73847 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58946 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22757 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56636 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15124 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74798 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46396 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61789 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37085 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43057 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19256 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21278 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64963 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19619 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77566 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18804 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64357 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61822 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64363 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15878 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12522 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6160 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46945 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1724 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48016 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47758 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
     return extractTargets(WTFMove(nodes), WTFMove(innerElement), request.canIncludeNearbyElements);
 }
 
+void ElementTargetingController::topologicallySortElementsHelper(WebCore::ElementIdentifier currentElementID, Vector<WebCore::ElementIdentifier> &depthSortedIDs, HashSet<WebCore::ElementIdentifier> &processingIDs, HashSet<WebCore::ElementIdentifier> &unprocessedIDs, const HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> &elementIDToOccludedElementIDs)
```
```suggestion
void ElementTargetingController::topologicallySortElementsHelper(ElementIdentifier currentElementID, Vector<ElementIdentifier>& depthSortedIDs, HashSet<ElementIdentifier>& processingIDs, HashSet<ElementIdentifier>& unprocessedIDs, const HashMap<ElementIdentifier, HashSet<ElementIdentifier>>& elementIDToOccludedElementIDs)
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    depthSortedIDs.append(elementID);
+}
+
+Vector<WebCore::ElementIdentifier> ElementTargetingController::topologicallySortElements(const HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> &elementIDToOccludedElementIDs)
```
```suggestion
Vector<ElementIdentifier> ElementTargetingController::topologicallySortElements(const HashMap<ElementIdentifier, HashSet<ElementIdentifier>>& elementIDToOccludedElementIDs)
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+{
+    Vector<WebCore::ElementIdentifier> depthSortedIDs;
+    HashSet<WebCore::ElementIdentifier> processingIDs;
+    HashSet<WebCore::ElementIdentifier> unprocessedIDs;
```
```suggestion
    Vector<ElementIdentifier> depthSortedIDs;
    HashSet<ElementIdentifier> processingIDs;
    HashSet<ElementIdentifier> unprocessedIDs;
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    return depthSortedIDs;
+}
+
+Vector<Vector<TargetedElementInfo>> ElementTargetingController::findAllTargets(WebCore::FloatSize minimumSize)
```
```suggestion
Vector<Vector<TargetedElementInfo>> ElementTargetingController::findAllTargets(FloatSize minimumSize)
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    for (auto x = viewportRect.x() + std::floor(hitTestRegionWidth / 2); x < viewportRect.maxX(); x += hitTestRegionWidth) {
+        for (auto y = viewportRect.y() + std::floor(hitTestRegionHeight / 2); y < viewportRect.maxY(); y += hitTestRegionHeight) {
+            const FloatPoint hitTestPoint = FloatPoint(x, y);
+            auto [nodes, innerElement] = findNodes(hitTestPoint, true);
```
```suggestion
            auto [nodes, innerElement] = findNodes(FloatPoint { x, y }, true);
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    }
+
+    HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> elementIDToOccludedElementIDs;
+    HashMap<WebCore::ElementIdentifier, Vector<TargetedElementInfo>> elementIDToTargets;
```
```suggestion
    HashMap<ElementIdentifier, HashSet<ElementIdentifier>> elementIDToOccludedElementIDs;
    HashMap<ElementIdentifier, Vector<TargetedElementInfo>> elementIDToTargets;
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+            continue;
+
+        const auto topElementID = targets.first().elementIdentifier;
+        HashSet<WebCore::ElementIdentifier> occludedElementIDsToInsert;
```
```suggestion
        HashSet<ElementIdentifier> occludedElementIDsToInsert;
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    HashMap<WebCore::ElementIdentifier, HashSet<WebCore::ElementIdentifier>> elementIDToOccludedElementIDs;
+    HashMap<WebCore::ElementIdentifier, Vector<TargetedElementInfo>> elementIDToTargets;
+    for (auto& targets : targetsList) {
+        if (targets.size() <= 0)
```
```suggestion
        if (targets.isEmpty())
```

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+            elementIDToTargets.set(topElementID, targets);
+
+        if (elementIDToOccludedElementIDs.contains(topElementID))
+            elementIDToOccludedElementIDs.set(topElementID,  elementIDToOccludedElementIDs.get(topElementID).unionWith(occludedElementIDsToInsert));
```
```suggestion
            elementIDToOccludedElementIDs.set(topElementID, elementIDToOccludedElementIDs.get(topElementID).unionWith(occludedElementIDsToInsert));
```

whsieh commented on Source/WebCore/page/ElementTargetingController.h:
```diff
@@ -56,6 +56,7 @@ class ElementTargetingController final : public CanMakeCheckedPtr<ElementTargeti
     ElementTargetingController(Page&);
 
     WEBCORE_EXPORT Vector<TargetedElementInfo> findTargets(TargetedElementRequest&&);
+    WEBCORE_EXPORT Vector<Vector<TargetedElementInfo>> findAllTargets(WebCore::FloatSize);
```
```suggestion
    WEBCORE_EXPORT Vector<Vector<TargetedElementInfo>> findAllTargets(FloatSize);
```

whsieh reviewed (commented):
Just some minor editorial comments — continuing to look...

Is it possible write a simple test (layout test or API test) for this? (Some of the other tests in either `ElementTargetingTests.mm` or `fast/element-targeting` might be helpful to consult)

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+        topologicallySortElementsHelper(*unprocessedIDs.begin(), depthSortedIDs, processingIDs, unprocessedIDs, elementIDToOccludedElementIDs);
+    }
+
+    depthSortedIDs.reverse();
+    return depthSortedIDs;
```
Just musing; nothing actionable: we could consider doing the collection agnostic 
`*std::ranges::reverse(depthSortedIDs);` in Cocoa code

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+            if (isFirstElement) {
+                isFirstElement = false;
+                continue;
+            }
```
This is not being used anywhere? Safe to remove?

aspensantos replied:
I'm using it to skip adding the first element in targets since that is the key in the dict, so it would create a cycle in the graph. lmk if there's a cleaner way to skip the first element!

aprotyas replied:
This is fine! I would've suggested `for (auto& target : std::ranges::drop_view { targets, 1 })` but this is WebCore code that needs to build on platforms that don't have great STL support.

aprotyas replied:
Actually why not just do an index/iterator-based loop that starts from the element 1? Might be easier to follow than the `isFirstElement` logic?

```
// Starting from index 1 to break a lookup cycle.
for (auto index = 1; index < targets.size(); ++target) {
    auto& target = targets.at(index);
    ...
}
```

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+        }
+
+        if (elementIDToTargets.contains(topElementID)) {
+            const auto storedIDsSet = elementIDToOccludedElementIDs.get(topElementID);
```
Does presence in `elementIDToTargets` necessarily imply presence in `elementIDToOccludedElementIDs`? Maybe there's an opportunity to codify that relation then.

aspensantos replied:
Updated this to combine the two sets of `if { ... } else { ... }` since if the element is in one it should also be in the other

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+
+        if (elementIDToTargets.contains(topElementID)) {
+            const auto storedIDsSet = elementIDToOccludedElementIDs.get(topElementID);
+            auto storedTargets = elementIDToTargets.get(topElementID);
```
Nit; consider
```suggestion
        if (auto storedTargets = elementIDToTargets.getOptional(topElementID)) {
            const auto storedIDs = elementIDToOccludedElementIDs.get(topElementID);
```

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,129 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    for (auto& elementID : topologicallySortedIDs)
+        topologicallySortedTargetsList.append(elementIDToTargets.get(elementID));
+
+    return topologicallySortedTargetsList;
```
Optional; this plumbing can be wrapped into a `map` operation. I don't think it harms legibility.

```suggestion
    return topologicallySortElements(elementIDToOccludedElementIDs).map([& elementIDToTargets](const auto& elementIdentifier) {
        return elementIDToTargets.get(elementIdentifier);
    });
```

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -569,6 +569,8 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
 
 - (void)_requestTargetedElementInfo:(_WKTargetedElementRequest *)request completionHandler:(void(^)(NSArray<_WKTargetedElementInfo *> *))completionHandler WK_API_AVAILABLE(macos(15.0), ios(18.0), visionos(2.0));
 
+- (void)_requestAllTargetableElementsInfo:(CGSize)minimumSize completionHandler:(void(^)(NSArray<NSArray<_WKTargetedElementInfo *> *> *))completionHandler WK_API_AVAILABLE(visionos(2.0));
```
You will need to provide availability macros for other platforms too, I believe.

aspensantos replied:
@whsieh said that since we only need this for visionOS we should avoid making it available for other platforms, do I need to specify this as unavailable on other platforms?

whsieh replied:
Oh, that's probably a good idea. You could also guard this behind a compile-time flag for visionOS.

whsieh replied:
Just a minor naming nit — the `minimumSize` argument feels a tiny bit unclear here. At first, I'd assumed it was a minimum size threshold for _hit-tested targets_, but then realized it was the size of the intervals between hit-testing locations when I read the implementation.

Not immediately obvious how to many this name clearer (without being overly verbose), but perhaps something along the lines of `hitTestInterval`?

aspensantos replied:
Marked them as unavailable but didn't compile them out so it would be easier to catch build errors that could be introduced by changing the names of methods used by this new API when developing on other platforms. Also updated the parameter to `hitTestInterval` ✅

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15146,6 +15146,24 @@ void WebPageProxy::requestTargetedElement(const API::TargetedElementRequest& req
     });
 }
 
+void WebPageProxy::requestAllTargetableElements(WebCore::FloatSize minimumSize, CompletionHandler<void(const Vector<Vector<Ref<API::TargetedElementInfo>>>&)>&& completion)
```
Make the completion handler argument an rvalue ref too? Since you're creating a new `Vector` anyway.

```suggestion
void WebPageProxy::requestAllTargetableElements(WebCore::FloatSize minimumSize, CompletionHandler<void(Vector<Vector<Ref<API::TargetedElementInfo>>>&&)>&& completion)
```

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15146,6 +15146,24 @@ void WebPageProxy::requestTargetedElement(const API::TargetedElementRequest& req
+            return completion({ });
+
+        completion(WTF::map(WTFMove(elements), [protectedThis](auto&& subelements) {
+            return WTF::map(WTFMove(subelements), [protectedThis](auto&& subelement) {
+                return API::TargetedElementInfo::create(*protectedThis, WTFMove(subelement));
```
`std::move(auto&&)` is ~ill-defined~ under-defined since this is a forwarding reference.


```suggestion
            return WTF::map(std::forward<decltype(subelements)>(subelements), [protectedThis](auto&& subelement) {
                return API::TargetedElementInfo::create(*protectedThis, std::forward<decltype(subelement)>(subelement));
```

aprotyas replied:
In hindsight, we aren't using the forwarding reference to bind to other reference types, so a `std::move` here is fine and sufficient. Excuse the pedantry :)

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9763,6 +9763,15 @@ void WebPage::requestTargetedElement(TargetedElementRequest&& request, Completio
+{
+    RefPtr page = corePage();
+    if (!page)
+        return completion({ { } });
```
I would've assumed `completion({ })` works to default initialize this vector.

aprotyas reviewed (commented):
Nice patch. Also, +1 to all of Wenson's suggestions.

whsieh commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -860,6 +861,123 @@ Vector<TargetedElementInfo> ElementTargetingController::findTargets(TargetedElem
+    if (processingIDs.contains(currentElementID)) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
```
Nit - I think you can skip 1 redundant hash lookup by rearranging the early returns:

```suggestion
    if (processingIDs.contains(currentElementID)) {
        ASSERT_NOT_REACHED();
        return;
    }

    if (!unprocessedIDs.contains(currentElementID))
        return;
```

aspensantos committed:
[Remote Inspection] Add the ability to fetch all targetable elements
https://bugs.webkit.org/show_bug.cgi?id=281173
rdar://135354335

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Add the ability to fetch all targetable elements to collect highlight regions for accessibility

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::topologicallySortElementsHelper):
(WebCore::ElementTargetingController::topologicallySortElements):
- Topologically sort elements represented in a directed graph

(WebCore::ElementTargetingController::findAllTargets):
- Find all the targets in view by using the hitTestInterval parameter to findTargets in sections of that size
- Store the z relationships between the targets found in the dictionary elementIDToOccludedElementIDs
- Use the topologicallySortElements function on elementIDToOccludedElementIDs to get the z sorted elementIDs
- Return the targets for each elementID in z sorted order

* Source/WebCore/page/ElementTargetingController.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestAllTargetableElementsInfo:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestAllTargetableElements):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestAllTargetableElements):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ElementTargetingTests.mm:
(-[WKWebView allTargetableElementsWithHitTestInterval:]):
(TestWebKitAPI::TEST(ElementTargeting, RequestAllVisibleElements)):
- Add API test for [WKWebView _requestAllTargetableElementsInfo:completionHandler:]

* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-11.html: Added.
- Add html file used in API test for [WKWebView _requestAllTargetableElementsInfo:completionHandler:]

Canonical link: https://commits.webkit.org/285200@main
